File: ui/src/main/kotlin/listenwithrandos/app/LogoutButton.kt
package listenwithrandos.app

import kotlinx.html.ButtonType.button
import kotlinx.html.js.onClickFunction
import listenwithrandos.state.AppAction
import listenwithrandos.state.LogoutAction
import react.*
import react.dom.button
import react.redux.rConnect
import redux.WrapperAction

interface LogoutButtonDispatchProps: RProps {
    var logOut: () -> Unit
}

interface LogoutButtonProps: LogoutButtonDispatchProps

class LogoutButton: RComponent<LogoutButtonProps, RState>() {
    override fun RBuilder.render() {
        button(type = button, classes = "btn btn-link nav-link") {
            attrs.onClickFunction = { props.logOut() }

            +"Log Out"
        }
    }
}

val connectedLogoutButton = rConnect<AppAction, WrapperAction, RProps, LogoutButtonDispatchProps>(
    mapDispatchToProps = { dispatch, _ ->
        logOut = { dispatch(LogoutAction) }
    }
)(LogoutButton::class.js.unsafeCast<RClass<RProps>>())

File: ui-new/src/app/Navbar/LogoutButton/LogoutButton.js
// @flow

import * as React from 'react'
import type {Dispatch} from '../../../domain/store/actions/actions'
import {LOGOUT} from '../../../domain/store/actions/actions'
import {connect} from 'react-redux'

type LogoutButtonDispatchProps = {
  logOut: () => any
}

const LogoutButtonComponent = (props: LogoutButtonDispatchProps) => (
  <button type="button" className="btn btn-link nav-link" onClick={props.logOut}>
    Log out
  </button>
)

const mapDispatchToProps = (dispatch: Dispatch) => ({
  logOut() {
    dispatch({type: LOGOUT})
  },
})

export const LogoutButton = connect(
  null,
  mapDispatchToProps,
)(LogoutButtonComponent)
